src/main/java/pages/Agoda/AgodaSearchResultsPage.java

```diff
@@ -99,12 +87,84 @@ public void verifyLowestPriceSortOrder(String expectedDestination, int expectedH
+        SelenideElement starRatingElement = $x(String.format("//div[@id='SideBarLocationFilters']//legend[@id='filter-menu-RecentFilters']//following-sibling::ul//li//label[@data-element-value='%s']//div//div//input", star));
+        starRatingElement.shouldBe(Condition.selected);
+        assertEquals(minPriceSlider.getAttribute("aria-valuetext"), formatVND(minPrice));
+        assertEquals(maxPriceSlider.getAttribute("aria-valuetext"), formatVND(maxPrice));
+    }
+
+    @Step("Verify filter results for destination: {0}, price range: {1} - {2}, stars: {3}, expected count: {4}")
+    public void verifyFilterResult(String expectedDestination, int minPrice, int maxPrice, String expectedStars, int expectedCount) {
+        hotelListings.shouldHave(CollectionCondition.sizeGreaterThanOrEqual(expectedCount));
+
+        int verifiedCount = 0;
+        int index = 0;
+
+        while (verifiedCount < expectedCount) {
+            if (index >= hotelListings.size()) {
+                throw new AssertionError("Expected at least " + expectedCount + " valid hotels, but found only " + verifiedCount);
+            }
 
+            SelenideElement hotelItem = hotelListings.get(index);
             index++;
+            if (isSoldOut(hotelItem)) continue;
+
+            verifyHotelDestination(hotelItem, expectedDestination);
+            verifyHotelPriceInRange(hotelItem, minPrice, maxPrice);
+            verifyHotelStars(hotelItem, expectedStars);
+
+            verifiedCount++;
         }
     }
 
+    @Step("Verify hotel destination matches expected: {1}")
+    public void verifyHotelDestination(SelenideElement hotelItem, String expectedDestination) {
+        SelenideElement hotelAreaElement = hotelItem.find(By.xpath(".//div[@data-selenium='area-city']"));
+        hotelAreaElement.scrollIntoView(true).shouldBe(Condition.visible);
+        String hotelAreaText = hotelAreaElement.getText();
+        assertTrue(hotelAreaText.toLowerCase().contains(expectedDestination.toLowerCase()), "Hotel destination mismatch: " + hotelAreaText);
+    }
+
+    @Step("Verify hotel price is within range: {1} - {2}")
+    public void verifyHotelPriceInRange(SelenideElement hotelItem, int minPrice, int maxPrice) {
+        double price = extractPrice(hotelItem);
+        assertTrue(price >= minPrice && price <= maxPrice, "Hotel price out of expected range: " + price);
+    }
+
+
+    @Step("Verify hotel stars match expected: {1}")
+    public void verifyHotelStars(SelenideElement hotelItem, String expectedStars) {
+        SelenideElement starElement = hotelItem.find(By.xpath(String.format(".//div[@data-testid='rating-container']//span[.='%s stars out of 5']", expectedStars)));
+        starElement.shouldBe(Condition.visible, defaultTimeout);
+        String actualStars = starElement.getText().split(" ")[0];
+        assertEquals(actualStars, expectedStars, "Star mismatch! Expected: " + expectedStars + ", got: " + actualStars);
```

Comment:
in case of assertion failed, testng already logged the expected and actual values for you